Paper: pulse2percept: A Python-based simulation framework for bionic vision

mbeyeler opened the pull request:
pulse2percept: A Python-based simulation framework for bionic vision
Michael Beyeler, Geoffrey M. Boynton, Ione Fine, Ariel Rokem

yingweiy commented:
I am reviewing this paper for Scipy. Here are my comments:

1. The paper is very well written and organized overall. It is technically sound.
2. How does the model/code get validated or tested? What are limitations (if any) with current implementation?
3. Page 7 - "Computational Performance": Do you have, or plan to have GPU implementation to further improve the performance?
4. Page 8, second paragraph "Ease of use": Since you mentioned that the users of this software most likely to be MDs instead of computer scientists. Have you considered to provide a GUI which may make the use easier for them?
5. As indicated in the requirements, the paper should not be longer than 8 pages. 
https://github.com/scipy-conference/scipy_proceedings/blob/master/review_criteria.md
 This paper is about 9 pages.

manojpandey commented on papers/michael_beyeler/michael_beyeler.rst:
```diff
@@ -0,0 +1,1210 @@
+
+   class MyGanglionCellModel(p2p.retina.TemporalModel):
+       def model_cascade(self, in_arr, pt_list, layers,
+                         dojit):
```
`do_jit` / `use_jit` would look more intuitive :)

mbeyeler replied:
We have renamed the variable `use_jit` in https://github.com/scipy-conference/scipy_proceedings/pull/275/commits/6ab45b0c84506df8eb9438a036036f771d143ce2. See also our detailed answer below.

manojpandey commented:
@katyhuff : I'm excited to review this paper by @mbeyeler for Scipy 2017 !

---
Wonderful write-up. `pulse2percept` shows lots of promises in the medicine field. 

Apart from what @yingweiy reviewed, I have left minor suggestions as review messages above !
Also from 
> Due to the current design constraints (see Discussion), the software is better suited for rapid prototyping rather than realtime execution.

Do you have any proposal on making the software better for realtime execution in the near future ? :)

Overall, I would love to see this paper getting published !

jbednar commented:
Looks good! Very clear, and a nice approach.  The underlying technology and library choices seem sound and appropriate, with tradeoffs identified and conveyed clearly.  It seems like a niche topic, but it also acts as an illustration of how to approach and solve problems using the Scientific Python ecosystem, which makes it useful as a case study as well.

dalippa commented:
Overall comments:
* Very well written overall.
* Paper is accessible, explaining enough details for novices while not being esoteric.
* Switches "person" mid way through. Be consistent with either first person plural ("we") or 3rd person narrative for clarity.
* Switches tense mid way through. Be consistent with either present or past tense.
* Would greatly benefit from a future work section

If this is presented as a talk or a poster, please consider a demonstration of this working live.

mbeyeler commented on papers/michael_beyeler/michael_beyeler.rst:
```diff
@@ -0,0 +1,1210 @@
+
+Linear responses in Fig. :ref:`figmodel` C, D, and F
+were modeled as temporal low-pass filters,
+or 'leaky integrators',
```
I'm afraid single-quotes are typeset 'like this' in reStructuredText, and LaTeX notation doesn't work (see https://www.python.org/dev/peps/pep-0012/#habits-to-avoid). I agree though that it looks kind of weird. I could change them to double-quotes if you like?

mpacer replied:
This has to do with use using rst to convert to LaTeX to convert to pdf. @gpoore worked on this for inside code blocks, do you have thoughts about doing it with the standard text as well?

gpoore replied:
@mpacer What I did in code blocks was disabling the backtick and single quote from being converted into curly quotes. In standard text, there isn't a way to enable this, because the backticks are already used for links and inline code, so there wouldn't be a way to distinguish backticks used for LaTeX quotes. However, it *is* possible to enter the Unicode curly quote characters directly (`“ ” ‘ ’`), and they come through fine. Apparently docutils also has a smart quotes option, but then you have to worry about the possibility of the algorithm breaking.

mbeyeler replied:
@gpoore OK that works, I can use Unicode curly quotes instead. Seems like the easiest workaround. Thanks!

mbeyeler commented:
Thank you @yingweiy, @manojpandey, @jbednar, and @dalippa for your thoughtful reviews! Please see our detailed responses below.

@yingweiy
---------

> How does the model/code get validated or tested? What are limitations (if any) with current 
> implementation?

*pulse2percept* comes with its own test suite, which can be run using `py.test`. We have added a note in Section "Software availability and development" that gives the exact command.

> Page 7 - "Computational Performance": Do you have, or plan to have GPU implementation to further 
> improve the performance?

We do! Although the current state of the software is better suited for
rapid prototyping due to the design constraints mentioned in the Discussion,
we plan to make the software more amenable to real-time execution
through GPU parallelization (via CUDA and Dask) as well as cluster computing
(via Spark).

We have added this remark to the Computational Performance section.

> Page 8, second paragraph "Ease of use": Since you mentioned that the users of this software most l
> likely to be MDs instead of computer scientists. Have you considered to provide a GUI which may 
> make the use easier for them?

Excellent point. In the future, we plan for our code to become the back end of a web application similar to [Keshavan et al., in press](https://doi.org/10.1016/j.neuroimage.2017.03.055).
The web application could be used to configure the various parameters of a simulation,
and then sends requests to the server, which in turn launches a simulation on the back-end.
We have added a note about these plans in the *Ease of use* bullet point.

> As indicated in the requirements, the paper should not be longer than 8 pages.
> https://github.com/scipy-conference/scipy_proceedings/blob/master/review_criteria.md
> This paper is about 9 pages.

The latest build fits within 8 pages.



@manojpandey
------------

> @katyhuff : I'm excited to review this paper by @mbeyeler for Scipy 2017 !
> Wonderful write-up. pulse2percept shows lots of promises in the medicine field.

Thank you for your kind words!

> Apart from what @yingweiy reviewed, I have left minor suggestions as review messages above !

I have responded to them directly (please see above).

> Also from
> > Due to the current design constraints (see Discussion), the software is better suited for rapid prototyping rather 
> > than realtime execution.
> Do you have any proposal on making the software better for realtime execution in the near future ? :)

We do! Although the current state of the software is better suited for
rapid prototyping due to the design constraints mentioned in the Discussion,
we plan to make the software more amenable to real-time execution
through GPU parallelization (via CUDA and Dask) as well as cluster computing
(via Spark).

We have added this remark to the Computational Performance section.



@jbednar
--------

> Looks good! Very clear, and a nice approach. The underlying technology and 
> library choices seem sound and appropriate, with tradeoffs identified and
> conveyed clearly. It seems like a niche topic, but it also acts as an
> illustration of how to approach and solve problems using the Scientific Python
> ecosystem, which makes it useful as a case study as well.

Thanks! We, too, are hoping it will be of general interest to the SciPy community as a case study.


@dalippa
--------

> - Very well written overall.
> - Paper is accessible, explaining enough details for novices while not being
>   esoteric.

Thanks!

> - Switches "person" mid way through. Be consistent with either first person plural ("we") or 3rd person narrative for clarity.

We noticed most-and-for-all that the Implementation and Results section
is inconsistent when it comes to first vs. third person
(e.g., "where you specify", "can be specified", "we can choose", etc.).
We have thus reworked the section and the rest of the paper
to consistently use the royal "we".

Let us know if you had other places in mind where this needs attention.

> - Switches tense mid way through. Be consistent with either present or past tense.

We noticed a switch in tense starting with the Computational Model of Bionic Vision
section. We have updated this section, as well as Implementation and Results,
to consisently use present tense.

Let us know if you had other places in mind where this needs attention.

> - Would greatly benefit from a future work section

We're almost out of space, so instead of adding a separate section,
we opted to add remarks about our future plans in the following places:
- re: GPU and cluster computing in the Computational Performance section
- re: web application with Python back-end in the Discussion
- re: evaluating different electrode geometries and finding improved
  stimulation protocols in the Discussion

> - If this is presented as a talk or a poster, please consider a demonstration of this working live.

Sure thing!
I am planning on having a live demo during my talk Wednesday morning in the Neuroscience session.

dalippa commented on papers/michael_beyeler/michael_beyeler.rst:
```diff
@@ -0,0 +1,1238 @@
+~~~~~~~~~~~~~~~~~
+
+The *pulse2percept* project seeks a trade-off between 
+object-oriented programming and ease of use.
```
You may wish to explain why this is a tradeoff; for most professional programmers, good OO design *implies* ease of use.

mbeyeler replied:
Good point. The sentence was supposed to talk about the trade-off between optimizing for computational performance and ease of use.

dalippa commented on papers/michael_beyeler/michael_beyeler.rst:
```diff
@@ -0,0 +1,1238 @@
+
+*pulse2percept* offers a collection of functions to convert the 
+:code:`p2p.utils.TimeSeries` output into a movie file 
+(via Scikit-Video and ffmpeg).
```
no parentheses required; "using Scikit-Video and ffmpeg". Citations would also be good.

dalippa commented on papers/michael_beyeler/michael_beyeler.rst:
Consider including license information as well. Possibly earlier in the paper (e.e.g "a BSD-licensed Python package") with a reference to `https://github.com/uwescience/pulse2percept/blob/master/LICENSE`

mbeyeler commented on papers/michael_beyeler/michael_beyeler.rst:
Great suggestion! I added the note and ref on first mention.

mbeyeler commented:
The latest commit should have addressed all remaining concerns, specifically:

- Rephrase the sentence about trade-off computation performance <=> ease of use
- Add BSD license ref
- Add refs to scikit-image, scikit-video, and ffmpeg

dalippa commented:
@katyhuff @mpacer @deniederhut @mbeyeler Paper is ready for merging.
